### PR TITLE
refactor(transport): split into transport-common / transport / transport-web

### PR DESCRIPTION
**Pull request by mroz22**
## Summary

Split `@trezor/transport` along the environment boundary into three packages so that node consumers stop pulling DOM USB types and browser consumers stop pulling node `usb`/`dgram`:

- **`@trezor/transport-common`** — env-agnostic abstractions: `AbstractTransport`, `AbstractApiTransport`, `UsbApi`, structural `UsbInterfaceApi`/`UsbDeviceLike`, `SessionsBackground`/`SessionsClient`, THP, errors, constants, types, utils.
- **`@trezor/transport`** (slimmed, node-only) — `BridgeTransport`, `NodeUsbTransport`, `UdpTransport`, `UdpApi`, `createBridgeTransports`, pinger.
- **`@trezor/transport-web`** — `WebUsbTransport`, `BrowserSessionsBackground`, sessions shared-worker entry.

> **Relationship to #27110:** this PR is the package-split slice carved out of the broader transport refactor in **#27110**. Once this lands, #27110 is rebased on top and reduces to the pure-DI + composition-root registry changes. The pure-DI realization is stacked on this branch as **#27842**.

## Implementation notes

- Ambient `@types/w3c-web-usb` is replaced with structural `UsbInterfaceApi`/`UsbDeviceLike`/`UsbInTransferResultLike` shapes so abstract types no longer leak DOM globals.
- All three packages are marked `sideEffects: false` so webpack tree-shakes the unused implementations.
- Every consumer is updated: `@trezor/connect`, `@suite/*`, `suite-desktop-core`, `suite-native`, `connect-cli`, `transport-bridge`, `transport-bluetooth`, `transport-native-{usb,bluetooth}`, `transport-test`.
- Browser/react-native stubs for `nodeusb` and `udp` (via `package.json` `browser`/`react-native` redirects) are kept. `@trezor/connect/src/device/TransportList.ts` still statically imports all transport classes, so the stubs are required to prevent `usb` (top-level CJS side effects) and `dgram` from leaking into web/react-native bundles. Static-import removal is the pure-DI follow-up tracked in #27110.

## Migration notes for consumers

- Abstract types/utilities: `@trezor/transport` → `@trezor/transport-common` (`AbstractTransport`, `Transport`, `TRANSPORT`, `TRANSPORT_ERROR`, `UsbApi`, `SessionsBackground`/`SessionsClient`, etc.)
- Node concrete transports stay in `@trezor/transport` (`BridgeTransport`, `NodeUsbTransport`, `UdpTransport`, `createBridgeTransports`).
- `WebUsbTransport` moves to `@trezor/transport-web`.

This PR intentionally preserves the existing pure-string transport API in `@trezor/connect` (`'BridgeTransport'`/`'WebUsbTransport'`/etc.). Pure-DI and the suite composition-root transport registry are the follow-up tracked in #27110.

## Test plan

- [ ] CI green (build-app/desktop, build-web, transport-test, connect-test, lint, depcheck, type-references)
- [ ] Manual: suite-web bundle size sanity (no `dgram`/`usb` in the browser chunk)
- [ ] Manual: suite-desktop e2e smoke (USB + bridge transport)

## 🔍 Currents Test Results

🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=mroz22/transport-package-split&lookback=7)

🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=mroz22/transport-package-split&lookback=7)

🔍 **Suite native android test results:** [View in Currents](https://app.currents.dev/projects/iUe1Y4/runs?branch=mroz22/transport-package-split&lookback=7)

## 🔒 Quarantined E2E Tests

<details><summary>Trezor Suite (web) — 4 test(s)</summary>

| Test | Type |
|------|------|
| Quarantine test: "Onboarding - create wallet,Success (basic)" | 🙋 manual |
| Quarantine test: "Database migration,Db migration between: release/22.5/web => develop/web" | 🙋 manual |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |
| Quarantine test: "Recovery T2T1 - dry run,Recovery after partial recovery" | 🙋 manual |

</details>

_Updated: 2026-05-19T13:13:46.219Z • 4 test(s) total_

<details><summary>Trezor Suite (desktop) — 11 test(s)</summary>

| Test | Type |
|------|------|
| Account types suite > Add-account-types-non-BTC-coins | 🤖 auto |
| Onboarding - create wallet > Success (basic) | 🤖 auto |
| Quarantine test: "Receive transaction,Receive a ETH transaction" | 🙋 manual |
| Quarantine test: "Receive transaction,Receive a ETH transaction" | 🙋 manual |
| Quarantine test: "Global receive and send,Global receive" | 🙋 manual |
| Bridge > App acquired device, EXTERNAL bridge is restarted, app reconnects | 🤖 auto |
| Bridge > App spawns bundled bridge and stops it after app quit | 🤖 auto |
| Bridge > App in daemon mode spawns node-bridge | 🤖 auto |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |
| Quarantine test: "Send Base,User can perform ethereum sending on base network" | 🙋 manual |
| Quarantine CANARY test: "Use regtest to test pending transactions,Send couple of pending txs and check that they are pending until mined" | 🙋 manual |

</details>

_Updated: 2026-05-06T09:50:35.036Z • 11 test(s) total_

## 🌐 Preview deployments

🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/mroz22/transport-package-split/web/](https://dev.suite.sldev.cz/suite-web/mroz22/transport-package-split/web/)

### CODE REVIEW

File: packages/transport/src/transports/bridge.ts [hunk 1]
```diff
@@ -5,31 +5,29 @@ import {
...
+    TRANSPORT,
+} from '@trezor/transport-abstract';
+import * as ERRORS from '@trezor/transport-abstract/src/errors';
+import { parseThpMessage } from '@trezor/transport-abstract/src/thp/receive';
+import { bridgeApiCall } from '@trezor/transport-abstract/src/utils/bridgeApiCall';
+import * as bridgeApiResult from '@trezor/transport-abstract/src/utils/bridgeApiResult';
+import { createProtocolMessage } from '@trezor/transport-abstract/src/utils/bridgeProtocolMessage';
+import { receiveAndParse } from '@trezor/transport-abstract/src/utils/receive';
+import { error, success } from '@trezor/transport-abstract/src/utils/result';
+import { buildMessage } from '@trezor/transport-abstract/src/utils/send';
```
[the remark this answers is not in the document]
Comment (mroz22) [reply packages/transport/src/transports/bridge.ts]: Not an issue: the publish pipeline rewrites these. `scripts/publish/babel-plugin-sanitize-internal-imports.js` runs over the emitted JS (via `replace-imports.sh`) and turns `@trezor/<pkg>/src/...` into `@trezor/<pkg>/lib/...` (CJS) or `@trezor/<pkg>/libESM/...` (ESM). `@trezor/transport-abstract`'s `publishConfig.exports` already exposes the corresponding `"./lib/*"` and `"./libESM/*"` patterns, so the rewritten paths resolve.

Verified locally on the built artefact:

```
$ grep '@trezor/transport-abstract' packages/transport/lib/transports/bridge.js
const transport_abstract_1 = require("@trezor/transport-abstract");
const ERRORS = tslib_1.__importStar(require("@trezor/transport-abstract/lib/errors"));
const receive_1 = require("@trezor/transport-abstract/lib/thp/receive");
const bridgeApiCall_1 = require("@trezor/transport-abstract/lib/utils/bridgeApiCall");
...
```

Same convention is already in use on `develop` for `@trezor/transport/src/...` deep imports across the codebase (see e.g. `packages/transport-bridge/src/core.ts` on develop) and ships fine.

File: packages/connect/src/device/DeviceCurrentSession.ts [hunk 2]
```diff
@@ -5,9 +5,13 @@ import { ERRORS } from '@trezor/connect-common/src/constants';
...
-import type { MessageResponse, Session, Transport } from '@trezor/transport';
-import { TRANSPORT } from '@trezor/transport';
-import { isErrorWithoutDeviceInteraction } from '@trezor/transport/src/errors-groups';
+import {
+    type MessageResponse,
+    type Session,
+    TRANSPORT,
+    type Transport,
+} from '@trezor/transport-abstract';
+import { isErrorWithoutDeviceInteraction } from '@trezor/transport-abstract/src/errors-groups';
```
[the remark this answers is not in the document]
Comment (mroz22) [reply packages/connect/src/device/DeviceCurrentSession.ts]: Not an issue: the publish pipeline rewrites these. `scripts/publish/babel-plugin-sanitize-internal-imports.js` runs over the emitted JS (via `replace-imports.sh`) and turns `@trezor/<pkg>/src/...` into `@trezor/<pkg>/lib/...` (CJS) or `@trezor/<pkg>/libESM/...` (ESM). `@trezor/transport-abstract`'s `publishConfig.exports` already exposes the corresponding `"./lib/*"` and `"./libESM/*"` patterns, so the rewritten paths resolve.

Verified locally on the built artefact:

```
$ grep '@trezor/transport-abstract' packages/transport/lib/transports/bridge.js
const transport_abstract_1 = require("@trezor/transport-abstract");
const ERRORS = tslib_1.__importStar(require("@trezor/transport-abstract/lib/errors"));
const receive_1 = require("@trezor/transport-abstract/lib/thp/receive");
const bridgeApiCall_1 = require("@trezor/transport-abstract/lib/utils/bridgeApiCall");
...
```

Same convention is already in use on `develop` for `@trezor/transport/src/...` deep imports across the codebase (see e.g. `packages/transport-bridge/src/core.ts` on develop) and ships fine.

File: packages/transport-bridge/src/core.ts [hunk 3]
```diff
@@ -7,24 +7,29 @@ import {
...
+import { SessionsClient } from '@trezor/transport-abstract/src/sessions/client';
+import {
+    callThpMessage,
+    receiveThpMessage,
+    sendThpMessage,
+} from '@trezor/transport-abstract/src/thp';
+import { createProtocolMessage } from '@trezor/transport-abstract/src/utils/bridgeProtocolMessage';
+import { receive as receiveUtil } from '@trezor/transport-abstract/src/utils/receive';
+import { error, success, unknownError } from '@trezor/transport-abstract/src/utils/result';
+import { createChunks, sendChunks } from '@trezor/transport-abstract/src/utils/send';
```
[the remark this answers is not in the document]
Comment (mroz22) [reply packages/transport-bridge/src/core.ts]: Not an issue: the publish pipeline rewrites these. `scripts/publish/babel-plugin-sanitize-internal-imports.js` runs over the emitted JS (via `replace-imports.sh`) and turns `@trezor/<pkg>/src/...` into `@trezor/<pkg>/lib/...` (CJS) or `@trezor/<pkg>/libESM/...` (ESM). `@trezor/transport-abstract`'s `publishConfig.exports` already exposes the corresponding `"./lib/*"` and `"./libESM/*"` patterns, so the rewritten paths resolve.

Verified locally on the built artefact:

```
$ grep '@trezor/transport-abstract' packages/transport/lib/transports/bridge.js
const transport_abstract_1 = require("@trezor/transport-abstract");
const ERRORS = tslib_1.__importStar(require("@trezor/transport-abstract/lib/errors"));
const receive_1 = require("@trezor/transport-abstract/lib/thp/receive");
const bridgeApiCall_1 = require("@trezor/transport-abstract/lib/utils/bridgeApiCall");
...
```

Same convention is already in use on `develop` for `@trezor/transport/src/...` deep imports across the codebase (see e.g. `packages/transport-bridge/src/core.ts` on develop) and ships fine.

**Comment by mroz22** [comment]
## Claude PR Journal

### 2026-05-04 — Rebase

**Action taken:** Rebased `mroz22/transport-package-split` onto current `origin/develop` (HEAD `9bc86a13`). The branch was 12 commits behind; rebase completed cleanly with no conflicts.

**E2E failures note:** All e2e shards (suite-web 1–5, suite-desktop 1–7, connect 1) are failing. Investigation indicates this is **not** a rebase issue — the failures are already present in the pre-rebase run and are inherent to the transport package split itself. Static checks (Lint, Type Checking, Other Checks, Build libs) all pass. The test plan in the PR description lists "CI green" as an unchecked item, consistent with these being known open issues. No automated fix was applied.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/connect/src/device/TransportList.ts [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
-    NodeUsbTransport,
-    UdpTransport,
-    WebUsbTransport,
+    type AbstractTransportParams,
+    type Transport,
     isTransportInstance,
-} from '@trezor/transport';
-import type { AbstractTransportParams } from '@trezor/transport/src/transports/abstract';
+} from '@trezor/transport-abstract';
+import { WebUsbTransport } from '@trezor/transport-web';
```
[the remark this answers is not in the document]
Comment (mroz22) [reply packages/connect/src/device/TransportList.ts:12]: Valid concern, and the immediate consequence (browser bundles still resolving the node-only transports through `@trezor/connect`) is exactly what was breaking CI here:

- `usb` (top-level CJS event listeners) leaked into the renderer and `vendors.js` failed at load with `ReferenceError: require is not defined` (suite-desktop e2e + connect-test web shards).
- Metro could not resolve `dgram` from `transport/src/api/udp.ts` (`prepare_android_test_app`).
- vitest browser mode hit a CJS interop edge case on `transport/mocks/usb.js` because the `usb` import was reaching the alias at all (`web-thpPairing` / `web-authorizeCoinjoin`).

For this PR I've restored the `package.json` `browser` / `react-native` redirects and the corresponding `nodeusb.browser.ts` / `udp.browser.ts` stubs (using the new `@trezor/transport-abstract` types) — see ebf95da2a3. That keeps the package split intact while preventing the node-only modules from being pulled into web/react-native graphs at the resolver level.

Removing the static imports in `TransportList.ts` (so the redirects are no longer needed) is part of the pure-DI / composition-root work in #27110, as called out in the PR description.

File: packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json [hunk 5]
```diff
@@ -8,9 +8,11 @@
         "@trezor/protobuf",
         "@trezor/protocol",
         "@trezor/schema-utils",
+        "@trezor/transport-abstract",
         "@trezor/type-utils",
         "@trezor/utils",
         "bignumber.js",
+        "cross-fetch",
```
Comment (mroz22) [inline packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json]: Why did cross-fetch become part of this snapshot after our changes? I would expect that splitting packages shoud limit dependencies spill, not increase it.

File: packages/suite-build/configs/base.webpack.config.ts [hunk 6]
```diff
@@ -55,6 +55,7 @@ const config: webpack.Configuration = {
             events: require.resolve('events'),
             // Not required
             child_process: false,
+            dgram: false,
```
Comment (mroz22) [inline packages/suite-build/configs/base.webpack.config.ts]: Why did dgram become part of thisc config after our changes? I would expect that splitting packages shoud limit dependency spills, not increase it.

File: packages/transport-common/src/types/usbInterface.ts
Comment (mroz22) [inline packages/transport-common/src/types/usbInterface.ts:1]: can't these types be taken from the node 'usb' dependency?

File: packages/transport-common/.depcheckrc.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+    "ignore-patterns": ["libDev", "lib"],
+    "ignores": [
+        "@types/sharedworker",
+        "@types/bytebuffer",
+        "@babel/preset-env",
+        "@babel/preset-typescript"
```
Comment (mroz22) [inline packages/transport-common/.depcheckrc.json]: why are there new ignores, specifically  "@types/sharedworker" and "@types/bytebuffer"?

File: packages/transport-test/e2e/api/api.test.ts [hunk 8]
```diff
@@ -11,7 +11,7 @@ const setupApisUnderTest = async () => {
 
     if (typeof window !== 'undefined' && 'usb' in window.navigator) {
         window.Buffer = (await import('buffer')).Buffer;
-        usbInterface = window.navigator.usb;
+        usbInterface = (window.navigator as unknown as { usb: UsbInterfaceApi }).usb;
```
Comment (mroz22) [inline packages/transport-test/e2e/api/api.test.ts:21]: hmm why this cast is needed?

File: packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json (hunk 5 above)
Comment (mroz22) [reply packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json]: Good catch. This is a package-level transitive dependency effect, not something connect-mobile actively needs at runtime.

The new path is:

`@trezor/connect-mobile` -> `@trezor/connect-common` -> `@trezor/transport-abstract` -> `cross-fetch`

`cross-fetch` is there because `bridgeApiCall` moved into `transport-abstract` together with the shared bridge utilities. That made the helper reusable from both `@trezor/transport` and `@trezor/transport-bridge`, but it also means every abstract consumer now sees the HTTP bridge dependency in this conservative public-dependency snapshot.

So yes, this is a dependency spill. The cleaner split would be to keep bridge HTTP helpers out of `transport-abstract` and move them back to the node/bridge side, or into a smaller bridge-specific shared module/subpath, so mobile/common consumers do not inherit `cross-fetch` just because they need abstract transport types/constants.

File: packages/suite-build/configs/base.webpack.config.ts (hunk 6 above)
Comment (mroz22) [reply packages/suite-build/configs/base.webpack.config.ts]: This is also caused by the split being only the package-boundary step, not the full DI/composition-root cleanup yet.

`@trezor/connect/src/device/TransportList.ts` still statically imports `UdpTransport` from the node transport package. The browser/react-native redirects added in this PR make that import resolve to the stub in those environments, but webpack can still encounter the node-only `dgram` path depending on resolution order/config. `dgram: false` does not include or polyfill `dgram`; it explicitly tells webpack to ignore it instead of failing the browser build.

So I agree with the direction: this should be temporary. Once #27110 removes the static node transport imports from `TransportList.ts`, this fallback should be removable as well.

File: packages/transport-common/src/types/usbInterface.ts
Comment (mroz22) [reply packages/transport-common/src/types/usbInterface.ts:1]: I would avoid taking these from the node `usb` package in `transport-abstract`.

The `usb` package is a node/native dependency, while `transport-abstract` is meant to be environment-agnostic and consumed by browser, node, native, and common packages. Its WebUSB-facing types also reference the ambient WebUSB globals from `@types/w3c-web-usb` (`USB`, `USBDevice`, `USBConnectionEvent`, `BufferSource`, etc.), which is exactly what we wanted to avoid leaking into node/common consumers.

These local types are intentionally structural and limited to the subset `UsbApi` actually uses. Both browser `navigator.usb` and node `new WebUSB(...)` satisfy the shape, but the public `.d.ts` does not force consumers to load DOM/WebUSB ambient types or depend on the node `usb` package.

File: packages/transport-common/.depcheckrc.json (hunk 7 above)
Comment (mroz22) [reply packages/transport-common/.depcheckrc.json]: These were carried over from the old `packages/transport/.depcheckrc.json` while splitting the package.

After looking at the final package boundaries, `@types/sharedworker` belongs to `@trezor/transport-web`, where the SharedWorker entry lives. It does not look necessary in `transport-abstract` anymore. `@types/bytebuffer` also looks like legacy transport baggage now; I do not see a direct `ByteBuffer` usage in `transport-abstract` after the protobuf changes.

So the answer is: they are not intentionally new abstract-package requirements. They should probably be removed from `transport-abstract` rather than ignored there.

File: packages/transport-test/e2e/api/api.test.ts (hunk 8 above)
Comment (mroz22) [reply packages/transport-test/e2e/api/api.test.ts:21]: The cast bridges the browser global WebUSB type to our local structural `UsbInterfaceApi` type.

At runtime the guard checks that `window.navigator.usb` exists, but TypeScript does not reliably narrow `Navigator` to a shape with a `usb` property matching our local interface. We also intentionally do not want this test package path to rely on the ambient `@types/w3c-web-usb` `USB` type, because the new `UsbApi` contract is the local DOM-free structural type from `transport-abstract`.

So the cast says: in this browser-only branch, treat `navigator.usb` as the same structural interface that `UsbApi` accepts. The node branch still uses `new WebUSB(...)`, and both producers satisfy the same `UsbInterfaceApi` shape.

**Comment by mroz22** [comment]
NOTE: update-issue-comment unavailable in this session; new comment created instead of updating existing journal.

### Claude CI Journal — PR #27297

| Field | Value |
|---|---|
| Run date | 2026-05-10 |
| Trigger | Routine CI maintenance pass |

#### ⚠️ CI Failure — Human Attention Required

| Check | Status | Run |
|---|---|---|
| `build-deploy` | ❌ FAILED | [25538954518](https://github.com/trezor/trezor-suite/actions/runs/25538954518) |

**Classification:** NOT retryable — `build-deploy` completed in ~2.5 min (started 2026-05-09T08:34:22Z, ended 2026-05-09T08:37:02Z). Short duration indicates a deterministic build failure, not an infrastructure timeout.

**Note:** This PR (transport package split) has known e2e failures per the existing journal (2026-05-04). The `build-deploy` failure is a new separate issue.

**Action:** Manual investigation required. No auto-rerun possible.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/transport-web/.depcheckrc.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+    "ignore-patterns": ["libDev", "lib"],
+    "ignores": ["@types/sharedworker"]
```
Comment (peter-sanderson) [inline packages/transport-web/.depcheckrc.json:4]: this is sus, why? at least :pray: add explanation comment and link to issue to fix it

File: packages/transport-web/package.json [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    ],
+    "scripts": {
+        "depcheck": "yarn g:depcheck",
+        "type-check": "yarn g:tsc --build tsconfig.json",
+        "build:lib": "yarn g:rimraf ./libESM && yarn g:tsc --build tsconfig.libESM.json && ../../scripts/publish/replace-imports.sh ./libESM esm",
+        "lint:js": "yarn g:eslint '**/*.{ts,tsx,js}'"
+    },
+    "devDependencies": {
+        "@trezor/eslint": "workspace:*",
+        "@types/node": "22.13.10",
```
Comment (mroz22) [inline packages/transport-web/package.json]: why it still has "@types/node" ?

File: packages/transport-web/tsconfig.libESM.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+    "extends": "../../tsconfig.libESM.json",
+    "include": ["./src"],
+    "compilerOptions": {
+        "outDir": "./libESM",
+        "types": ["node"]
```
Comment (mroz22) [inline packages/transport-web/tsconfig.libESM.json]: the same question here? why "types": ["node"] ?

**Comment by mroz22** [comment]
NOTE: update-issue-comment unavailable; new comment added for updated failure set (supersedes 2026-05-10 entry covering build-deploy RUN 25538954518).

<details>
<summary>2026-05-15 — Non-retryable Other Checks failure (new run) ❌</summary>

**CI analysis (2026-05-15 routine pass):**

| Check | Status | Run | Classification |
|-------|--------|-----|----------------|
| `Other Checks` | ❌ FAILED | [25865190658](https://github.com/trezor/trezor-suite/actions/runs/25865190658) | NOT retryable — deterministic failure |

> ⚠️ **Human attention required** for `Other Checks` failure. All other checks (lint, type-check, unit tests, connect E2E, build) are green.

</details>

---
_Generated by [Claude Code](https://claude.ai/code)_

File: packages/transport-web/.depcheckrc.json (hunk 9 above)
Comment (mroz22) [reply packages/transport-web/.depcheckrc.json:4]: done

File: packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json (hunk 5 above)
Comment (szymonlesisz) [reply packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json]: +1 on this, i think we should keep `bridgeApiCall` and the rest of bridge related code in `@trezor/transport` for now. lets not not pollute `@trezor/transport-common` with  problematic dependencies.

maybe later we can even move whole bridge code to `@trezor/transport-bridge` package? (if we setup package to server-client modules similar to transport-bluetooth)

Comment (mroz22) [reply packages/requirements/src/requirements/public-package-dependencies/__snapshots__/connect-mobile.json]: ok, bridgeApiCall and related moved back to transport for now.

**Review by szymonlesisz — APPROVED** [review]
👍
